index.js

```diff
@@ -1,22 +1,71 @@
 'use strict'
 
+const { ServerResponse } = require('http')
 const fp = require('fastify-plugin')
 const websocket = require('websocket-stream')
+const findMyWay = require('find-my-way')
+
+const kWs = Symbol('ws')
 
 function fastifyWebsocket (fastify, opts, next) {
+  if (opts.handle && typeof opts.handle !== 'function') {
+    return next(new Error('invalid handle function'))
+  }
   const handle = opts.handle
+    ? (req, res) => opts.handle(req[kWs], req)
+    : (req, res) => { req[kWs].socket.close() }
+
   const options = Object.assign({ server: fastify.server }, opts.options)
 
-  if (typeof handle !== 'function') {
-    return next(new Error('invalid handle function'))
-  }
+  const router = findMyWay({
+    ignoreTrailingSlash: true,
+    defaultRoute: handle
+  })
 
-  const wss = websocket.createServer(options, handle)
+  const wss = websocket.createServer(options, handleRouting)
 
   fastify.decorate('websocketServer', wss)
 
+  fastify.addHook('onRoute', routeOptions => {
+    if (routeOptions.websocket || routeOptions.wsHandler) {
+      if (routeOptions.method !== 'GET') {
+        return next(new Error('websocket handler could be declared only in GET method'))
```

Comment:
Just a nit ^^
```suggestion
        return next(new Error('websocket handler can only be declared in GET method'))
```

Reply:
Actually i am having hard time testing this errors. I tried to use the same method as here https://github.com/fastify/fastify-websocket/blob/master/test.js#L139 but in my case err is null and all seems ok, even if the execution of the test enters in that if. Any suggestions?

Reply:
Actually you should just `throw` here.

Reply:
Ah ok. So, my first implementation was ok. 
Any docs to point me on?
(Sorry)

Reply:
Also here if I just throw i have an Avvio error: "ERR_AVVIO_PLUGIN_TIMEOUT".

Reply:
Then you might have caught a bug in Fastify :/. Would you mind opening a fix there? We are likely missing a try-catch wrapper on `'onRoute'`.

Reply:
OK! I'll do it ASAP.

Reply:
It was really a bug. Fixed with https://github.com/fastify/fastify/pull/1651